File: java/longest_substring_without_repeating_characters.java
public class Solution {
    public int lengthOfLongestSubstring(String s) {
        Map<Character, Integer> pos=new HashMap<>();
        int res=0;
        int start=0;
        for (int i=0; i<s.length(); i++) {
            if (pos.containsKey(s.charAt(i))) {
                int j=pos.get(s.charAt(i));
                res=Math.max(res, i-start);
                for (int k=start; k<=j; k++) {
                    pos.remove(s.charAt(k));
                }
                start=j+1;
            }
            
            pos.put(s.charAt(i), i);
        }
        res=Math.max(res, s.length()-start);
        
        return res;
    }
}

File: python/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        dic=collections.defaultdict(int)
        res, l, r, n=0, 0, 0, len(s)
        while r<n:
            if s[r] in dic:
                res=max(res, r-l)
                while l<=dic[s[r]]:
                    dic.pop(s[l])
                    l+=1
            dic[s[r]]=r
            r+=1
        res=max(res, r-l)
        return res

